### PR TITLE
docs: add background queue docs

### PR DESCRIPTION
**Pull request by levibostian**

### CODE REVIEW

File: docs/dev-notes/BACKGROUND-QUEUE.md [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+
+* Run the queue. 
+
+When it's time to run the queue, here are the operations of the queue:
+1. Take a snapshot of the queue inventory. Read the whole JSON file, grab the first task of the queue, execute that asynchronous operation. 
+2. If the task succeeds, delete the task from the file system and inventory,  move to the next task of the queue, go back to step 1 above. 
+3. If the task fails to run, the queue will check if the task that failed is the start of a *task group*. If it is *not* the start, the queue will simply behave as normal where it goes to the next task of the queue and executes it. if the task *is* the start of a task group, the queue will skip executing all members of that task group left in the queue. 
+
+What is a task group? Let's say the queue has these items in it:
+1. identify profile A - the API has never created profile A beore.
```
Comment (ami-aman) [inline docs/dev-notes/BACKGROUND-QUEUE.md:80]: Typo - *before

File: docs/dev-notes/BACKGROUND-QUEUE.md [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+2. track event for profile A
+3. push metric opened for push notification Z
+
+If queue task 1 fails to execute, queue task 2 will fail. Why? Because the API has yet to create profile A so we can't track events to profile A. To fix this issue, we created *task groups*. If we put task 1 and 2 in the same task group where task 1 is the *start* of the task group and have task 2 be a member of that same group, then if task 1 fails to execute, the queue will then see task 2 is a member of that group and will then skip task 2 and move onto task 3 which is not a member of the task group.
```
Comment (ami-aman) [inline docs/dev-notes/BACKGROUND-QUEUE.md:84]: Good. This is very well explained but I was wondering if it is possible to add a flowchart or an image to explain the same ?

File: docs/dev-notes/BACKGROUND-QUEUE.md [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+Here is an idea of the file system structure where the queue has 2 tasks in it:
+
+```
+Documents/
+  io.customer.ios/
+    queue/
+      inventory.json
+      tasks/
+        3838383949849493939393.json
+        2939929292001919202002.json
```
Comment (ami-aman) [inline docs/dev-notes/BACKGROUND-QUEUE.md:57]: Each task will have a separate JSON file ? Suppose out of 10 tasks , 2 tasks didn't execute successfully. Files related to those tasks will be deleted (and then recreated) or they will be there in the local file structure and removed only when tasks are done ?

File: docs/dev-notes/BACKGROUND-QUEUE.md [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+            // Handle failure. Determine what failed and why. 
+            // Write custom code to maybe retry the request later. 
+    }   
+}
+```
+
+With a queue, we can design our SDK to be similar to this:
+
+```swift 
+CustomerIO.shared.identify("Dana")
```
Comment (ami-aman) [inline docs/dev-notes/BACKGROUND-QUEUE.md:31]: This is powerful 🥇

**Comment by ami-aman** [comment]
I have implemented the updates with Background Queue so it was quick and easy for me to understand about BG Queues except few doubts which I have already listed down in the comments above. One last query related to BG queues :-

Background queue runs when either 10+ tasks are present in the queue OR after a timer of 30s. Right ? 

Now suppose, I am on workspace 12345, and I perform 3 activities on this workspace example, enable-disable-enable some habit, so the queue will 3 tasks. Queue will not run because either of both the conditions aren't met.  All good so far ! But I then change the workspace to my 56789 and redo some random activity like tap again on habits and new tasks are added to the queue. Now the queue reaches it's minimum 10 tasks condition and queue starts executing the tasks.

Question - Will the initial few tasks (which were created under workspace 12345) go to workspace 12345 or 56789 ?

File: docs/dev-notes/BACKGROUND-QUEUE.md (hunk 3 above)
Comment (levibostian) [reply docs/dev-notes/BACKGROUND-QUEUE.md:57]: Good question. I added a note about this to the doc. 

> After a task in the queue executes successfully, the queue will update the `inventory.json` file and will delete the task's `.json` file in `tasks/`. See `QueueStorage.swift` to view all the file system operations performed by the queue.

File: docs/dev-notes/BACKGROUND-QUEUE.md (hunk 2 above)
Comment (levibostian) [reply docs/dev-notes/BACKGROUND-QUEUE.md]: Great idea. I added a flowchart to the doc. Is it helpful?

Comment (ami-aman) [reply docs/dev-notes/BACKGROUND-QUEUE.md]: Looks perfect 👍

File: docs/dev-notes/BACKGROUND-QUEUE.md (hunk 3 above)
Comment (ami-aman) [reply docs/dev-notes/BACKGROUND-QUEUE.md:57]: Great. Thanks. 👍

**Comment by ami-aman** [comment]
> I have implemented the updates with Background Queue so it was quick and easy for me to understand about BG Queues except few doubts which I have already listed down in the comments above. One last query related to BG queues :-
> 
> Background queue runs when either 10+ tasks are present in the queue OR after a timer of 30s. Right ?
> 
> Now suppose, I am on workspace 12345, and I perform 3 activities on this workspace example, enable-disable-enable some habit, so the queue will 3 tasks. Queue will not run because either of both the conditions aren't met. All good so far ! But I then change the workspace to my 56789 and redo some random activity like tap again on habits and new tasks are added to the queue. Now the queue reaches it's minimum 10 tasks condition and queue starts executing the tasks.
> 
> Question - Will the initial few tasks (which were created under workspace 12345) go to workspace 12345 or 56789 ?

This question was discussed already in one of the team discussions

**Comment by ami-aman** [comment]
LGTM 👍

**Pull request merged**